module5/lessons/Week4/ProcessOptimization.md

```diff
@@ -0,0 +1,246 @@
+    // If a match is found, return the value
+```
+
+Another potential solution is to iterate over the array of numbers, and for each iteration, compare it to the rest of the array to see if a duplicate is found.
+</section>
+
+<section class="answer" markdown='1'>
+### Potential Solution C  
+
+**Hash/Obj Tracker**
+```
+// iterate over the array and add each value to a hash
+  // As soon as the value of a key is 2, you've found the duplicate
+```
+
+A third solution might include iterating over an array and adding it as a key to an object/hash, counting the number of instances in the value.  As soon as a key already exists and is increased to 2, you've found your duplicate.
+</section>
+
+### Choosing an Algorithm
+All of the options above will work to solve the problem, so how do we know which is 'best'?  We need some way of calculating or describing the time-complexity of an algorithm.  Typically, we describe time complexity in terms of how many steps it will take to complete in relation to an increasing number of elements.
+
+<section class='call-to-action' markdown='1'>
+
+In small groups, determine how many 'steps' are completed for the two code snippets below.
+* How many steps are represented by the current code?
+    * What if num[1] is a duplicate of num[0]?
+    * What if num[9] is a duplicate of num[5]?
+* How many steps would there be if `nums` had 25 elements?
+* How many steps would there be if `nums` had 50 elements?
+
+**Option 1**  
+```c#
+// find the duplicate
+int[] nums = {36, 84, 75, 20, 30, 44, 17, 98, 92, 27 };
+int? duplicate;
+
+foreach (var numA in nums)
+{
+    foreach (var numB in nums)
+    {
+        if (numA == numB)
+        {
+            duplicate = num1;
+        }
+    }
+}
+
+Console.WriteLine(duplicate);
+```
+**Option 2**
+```c#
+int[] nums = { 36, 84, 75, 20, 92, 44, 17, 98, 92, 27 };
+Dictionary<int, int> countOfNums = new Dictionary<int, int>();
+
+foreach (var num in nums)
+{
+    if (countOfNums.ContainsKey(num))
+    {
+        Console.WriteLine(num);
+        break;
+    }
+    countOfNums.Add(num, 1);
+}
+```
+</section>
+
+## Big O
+
+It's not very useful to have to describe algorithms in *exact* number of steps for every possible number of elements.  So, developers describe time complexity using mathematic expressions.  We call these expressions **Big O Notation** (big 'oh').  The BigO of an algorithm describes the _worst-case scenario_: what is the maximum number of stes an algorithm will take in relation to how the dataset grows.
+
+When describing BigO, we `"n"` to signify the amount of data. The most common examples you might see include:
+
+- O(1)
+- O(log n)
+- O(n)
+- O(n log n)
+- O(n^2) (aka, n-squared)
+
+We'll discuss each of these in order from least complex (best performance) to most complex (worst performance).
+
+<section class="answer" markdown='1'>
+### Constant Time - O(1)
+
+**Description:**
+- No matter how much data there is, it only takes one operation.
```

Comment:
I would probably change this wording to "No matter how much data there is, it takes the same number of operations". It often trips up students that something that always takes 3 operations is still O(1).

Reply:
totally - good call!